orchestrator/services/subscription_relations.py

```diff
@@ -0,0 +1,269 @@
+    """Function to efficiently get the depends_on SubscriptionTables for multiple subscription_ids."""
+    _filter_statuses: tuple[str, ...] = filter_statuses or tuple(SubscriptionLifecycle.values())
+    depends_on_relations = await _get_depends_on_relations(subscription_ids, _filter_statuses)
+
+    # Retrieve SubscriptionTable for all unique dependson ids
+    unique_depends_on_ids = {row.depends_on_sub_id for row in depends_on_relations}
+    _depends_on_subs = db.session.execute(
+        select(SubscriptionTable).filter(SubscriptionTable.subscription_id.in_(unique_depends_on_ids))
+    ).scalars()
+    depends_on_subs = {subscription.subscription_id: subscription for subscription in _depends_on_subs}
+
+    # group (more_itertools.bucket doesn't seem to work for tuple of uuids)
+    subscription_depends_on_ids: dict[UUID, list[UUID]] = {}
+    for relation in depends_on_relations:
+        subscription_depends_on_ids.setdefault(relation.in_use_by_sub_id, []).append(relation.depends_on_sub_id)
+
+    def get_depends_on_subs(in_use_by_id: UUID) -> list[SubscriptionTable]:
+        depends_on_ids = subscription_depends_on_ids.get(in_use_by_id, [])
+        return [depends_on_sub for id_ in depends_on_ids if (depends_on_sub := depends_on_subs.get(id_))]
+
+    # Important (as with any dataloader)
+    # Return the list of dependson subs in the exact same order as the ids passed to this function
+    return [get_depends_on_subs(subscription_id) for subscription_id in subscription_ids]
+
+
+async def get_recursive_relations(
+    subscription_ids: list[UUID],
+    filter_statuses: tuple[str, ...],
+    recurse_product_types: list[str],
+    recurse_depth_limit: int,
+    relation_fetcher: Callable[[list[UUID], tuple[str, ...]], Awaitable[list[list[SubscriptionTable]]]],
+    current_depth: int = 0,
+    used_subscription_ids: set[UUID] | None = None,
+) -> list[SubscriptionTable]:
+    """Recursively fetches subscription relations based on a custom relation-fetching function.
+
+    Args:
+        subscription_ids: List of subscription IDs to start with.
+        filter_statuses: List of statuses to filter the relations.
+        recurse_product_types: List of product types to recurse into.
+        recurse_depth_limit: Maximum recursion depth.
+        current_depth: Current recursion depth.
+        used_subscription_ids: List of already used subscription IDs to avoid cycles.
+        relation_fetcher: A function that fetches the relations for the subscriptions.
+
+    Returns:
+        A flattened list of all fetched relations for the given subscriptions.
+    """
+
+    used_subscription_ids = used_subscription_ids or set(subscription_ids)
+    relations = await relation_fetcher(subscription_ids, filter_statuses)
+    flat_relations_list = list(flatten(relations))
+
+    def get_related_subscription_ids() -> list[UUID]:
+        return [
+            r.subscription_id
+            for r in flat_relations_list
```

Comment:
Small suggestion to rename variables; the intermediate result and the fact that something is "flat" is not that relevant. 

Also since we can already get duplicates here, we can deduplicate them while we're iterating over the flattened result

```suggestion
    _relations = await relation_fetcher(subscription_ids, filter_statuses)
    relations = list(unique_everseen(flatten(_relations), key=lambda s: s.subscription_id))

    def get_related_subscription_ids() -> list[UUID]:
        return [
            r.subscription_id
            for r in relations
```